File: src/main/java/de/twometer/amongus/gui/CustomizePage.java
package de.twometer.amongus.gui;

import de.twometer.amongus.core.AmongUs;
import de.twometer.amongus.model.ClientSession;
import de.twometer.amongus.model.PlayerColor;
import de.twometer.amongus.model.SessionConfig;
import de.twometer.amongus.net.NetMessage;
import de.twometer.neko.util.Log;

public class CustomizePage extends BasePage {

    private SessionConfig config = new SessionConfig();
    private PlayerColor color = PlayerColor.Red;

    public CustomizePage() {
        super("Customize.html");
    }

    @Override
    public void onDomReady() {
        var i = 0;
        for (var color : PlayerColor.values()) {
            var col = color.getValue();
            var colStr = String.format("rgb(%d,%d,%d)", (int) (col.getR() * 255), (int) (col.getG() * 255), (int) (255 * col.getB()));
            context.call("setColorBox", i, colStr, color.name());
            i++;
        }
    }

    public void setConfig(String conf) {
        this.config = gson.fromJson(conf, SessionConfig.class);
        //Log.d("Config changed: " + config.getVotingTime() + " " + config.isConfirmEjects() + " " + config.getPlayerSpeed());
    }

    public void setColor(String color) {
        this.color = Enum.valueOf(PlayerColor.class, color);
        //Log.d("Set color to be " + color);
    }

    public void apply() {
        showLoading("Creating session...");
        // Create session
        amongUs.getClient().sendMessage(new NetMessage.SessionCreate())
                .await(NetMessage.SessionCreated.class, created -> {

                    if (created.result == NetMessage.SessionCreated.Result.MatchmakerFull) {
                        showError("The matchmaker is full. Please try again later.");
                        return;
                    }

                    // Join the session
                    showLoading("Joining...");
                    sendJoinMessage(created.code, () -> {
                        // Configure the session
                        showLoading("Configuring session...");
                        amongUs.getClient().sendMessage(new NetMessage.SessionConfigure(config))
                                .await(NetMessage.SessionConfigured.class, configured -> {

                                    // Configure the player
                                    showLoading("Configuring player...");
                                    amongUs.getClient().sendMessage(new NetMessage.ColorChange())
                                            .await(NetMessage.ColorChanged.class, changed -> {

                                                // Finally, let's go to the lobby
                                                AmongUs.get().getGuiManager().showPage(new LobbyPage());

                                            }).handleError(this::networkError);

                                }).handleError(this::networkError);

                    });

                }).handleError(this::networkError);
    }


    public void back() {
        goBack();
    }

}

File: src/main/kotlin/de/twometer/amogus/gui/CustomizePage.kt
package de.twometer.amogus.gui

import de.twometer.amogus.client.AmongUsClient
import de.twometer.amogus.model.PlayerColor
import de.twometer.amogus.model.SessionConfig
import de.twometer.amogus.net.ColorChangeRequest
import de.twometer.amogus.net.ColorChangeResponse
import de.twometer.amogus.net.SessionConfigureRequest
import de.twometer.amogus.net.SessionConfigureResponse

class CustomizePage : BasePage("Customize.html") {

    private var playerColor: PlayerColor? = null
    private var sessionConfig: SessionConfig? = null

    override fun onLoaded() {
        val values = PlayerColor.values()
        for (i in values.indices) {
            val value = values[i]
            val colorString = "rgb(${value.color.r * 255f}, ${value.color.g * 255f}, ${value.color.b * 255})"
            call("setColorBox", i, colorString, value.name)
            if (value == AmongUsClient.session!!.myself.color)
                call("selectColor", i)
        }
        call("setConfig", AmongUsClient.session!!.config)
        call("setIsHost", AmongUsClient.session!!.myselfIsHost)
        call("initialize")
    }

    fun setConfig(conf: String) {
        sessionConfig = conf.parseJson(SessionConfig::class.java)
    }

    fun setColor(color: String) {
        playerColor = PlayerColor.valueOf(color)
    }

    fun apply() {
        if (playerColor != null && playerColor != AmongUsClient.session!!.myself.color) {
            showLoading("Changing color...")
            AmongUsClient.send(ColorChangeRequest(playerColor!!))
            AmongUsClient.waitFor<ColorChangeResponse> {
                if (!it.accepted) {
                    showError("Color is already taken")
                } else applyConfig()
            }
        } else applyConfig()
    }

    private fun applyConfig() {
        if (!AmongUsClient.session!!.myselfIsHost || sessionConfig == null) {
            close()
            return
        }

        showLoading("Reconfiguring...")
        AmongUsClient.send(SessionConfigureRequest(sessionConfig!!))
        AmongUsClient.waitFor<SessionConfigureResponse> {
            if (!it.accepted)
                showError("Server rejected configuration")
            else
                close()
        }
    }

    fun back() = close()

}